cluster/restore-from-backup.sh

```diff
@@ -0,0 +1,188 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script performs disaster recovery of etcd from the backup data.
+# Assumptions:
+# - version.txt file is in the current directy (if it isn't it will be
+#     defaulted to "2.2.1/etcd2"). Based on this file, the script will
+#     decide to which version we are restoring (procedures are different
+#     for etcd2 and etcd3).
+# - in case of etcd2 - *.snap and *.wal files are in current directory
+# - in case of etcd3 - *.db file is in the current directory
+# - the script is run as root
+#
+# FIXME: The script currently doesn't support restoring of event etcd.
+
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# Version file contains information about current version in the format:
+# <etcd binary version>/<etcd api mode> (e.g. "3.0.12/etcd3").
+#
+# If the file doesn't exist we assume "2.2.1/etcd2" configuration is
+# the current one and create a file with such configuration.
+# The restore procedure is choosed based on this information.
+VERSION_FILE="version.txt"
+if [ ! -f "${VERSION_FILE}" ]; then
+  echo "2.2.1/etcd2" > "${VERSION_FILE}"
+fi
+VERSION_CONTENTS="$(cat ${VERSION_FILE})"
+ETCD_VERSION="$(echo $VERSION_CONTENTS | cut -d '/' -f 1)"
+ETCD_API="$(echo $VERSION_CONTENTS | cut -d '/' -f 2)"
+
+# Name is used only in case of etcd3 mode, to appropriate set the metadata
+# for the etcd data.
+# NOTE: NAME HAS TO BE EQUAL TO WHAT WE USE IN --name flag when starting etcd.
+NAME="${NAME:-etcd-$(hostname)}"
+
+# Wait until both etcd instances are up
+wait_for_etcd_up() {
+  health_ok="{\"health\": \"true\"}"
+  for i in $(seq 120); do
+    # TODO: Is it enough to look into /health endpoint?
+    etcd=$(curl --silent http://localhost:2379/health)
+    etcd_events=$(curl --silent http://localhost:4002/health)
+    if [ "${etcd}" == "${health_ok}" -a "${etcd-events}" == "${health_ok}" ]; then
+      return 0
+    fi
+    sleep 1
+  done
+  return 1
+}
+
+# Wait until apiserver is up.
+wait_for_apiserver_up() {
+  for i in $(seq 120); do
+    kubectl get componentstatuses 1>/dev/null 2>&1
+    if [ "$?" -eq "0" ]; then
+      return 0
+    fi
+    sleep 1
+  done
+  return 1
+}
+
+# Wait until etcd and apiserver pods are down.
+wait_for_etcd_and_apiserver_down() {
+  for i in $(seq 120); do
+    etcd=$(docker ps | grep etcd | grep -v etcd-empty-dir | wc -l)
+    apiserver=$(docker ps | grep apiserver | wc -l)
+    # TODO: Theoretically it is possible, that apiserver and or etcd
+    # are currently down, but Kubelet is now restarting them and they
+    # will reappear again. We should avoid it.
+    if [ "${etcd}" -eq "0" -a "${apiserver}" -eq "0" ]; then
+      return 0
+    fi
+    sleep 1
+  done
+  return 1
+}
+
+# Move the manifest files to stop etcd and kube-apiserver
+# while we swap the data out from under them.
+mv /etc/kubernetes/manifests/kube-apiserver.manifest ./
```

Comment:
What is our working directory when this runs?

Also, this strikes me as a little dangerous because there is nothing to restore it if the machine crashes at this point.